File: lib/read.js
// Reading the VPU mixer map off a device over AWJ. Server-side only.
//
// Path shape (firmware 6.2; verified against a real Aquilon C):
//
//   DeviceObject/preconfig/resources/{current|new}/status/mapping
//     /$device/@items/<1-4>                 1 = master, 2-4 = Link followers
//     /$vpuMixer/@items/PROC_<1-4>_MIXER_<1-16>
//       /@props/<prop>
//       /mixerAllocation/@props/usedOnOutPipe{1,2}
//       /$scaler/@items/{A,B}/@props/{memoryFill,memoryCut}
//
// The collection segment is `$vpuMixer`, camelCase. `$vpu-mixer` and `$mixer`
// both answer E12.

import { MIXER_IDS, MIXER_PROPS, SCALERS, OUT_PIPES } from '../public/vpu.js';

const base = (which, device, mixer) =>
  `DeviceObject/preconfig/resources/${which}/status/mapping` +
  `/$device/@items/${device}/$vpuMixer/@items/${mixer}`;

/**
 * Read one side of the mapping ("current" = running, "new" = staged).
 *
 * Mixers that are not fitted answer isAvailable=false; their detail reads are
 * skipped, which roughly halves the traffic on a part-populated chassis.
 *
 * Throws with code NO_VPU_SUBTREE if the very first probe finds no $vpuMixer
 * collection — that is a device without a VPU map (the simulator behaves this
 * way), and is worth telling the user plainly rather than showing an empty grid.
 *
 * @param {import('./awj.js').AwjClient} client
 * @param {{which:'current'|'new', device?:string|number}} opts
 */
export async function readMapping(client, { which, device = '1' } = {}) {
  const mixers = {};
  let first = true;

  for (const id of MIXER_IDS) {
    const b = base(which, device, id);
    const isAvailable = await client.tryGet(`${b}/@props/isAvailable`);

    if (isAvailable === undefined) {
      if (first) {
        const err = new Error(`no $vpuMixer collection under ${which} mapping`);
        err.code = 'NO_VPU_SUBTREE';
        throw err;
      }
      mixers[id] = { isAvailable: false };
    } else if (isAvailable !== true) {
      mixers[id] = { isAvailable: false };
    } else {
      const rec = { isAvailable: true };
      for (const p of MIXER_PROPS) rec[p] = await client.tryGet(`${b}/@props/${p}`);
      // There are EIGHT output pipes, one per output link — not two. Reading only
      // the first two hid three quarters of the device's own placement data.
      rec.mixerAllocation = {};
      for (let k = 1; k <= OUT_PIPES; k++) {
        rec.mixerAllocation[`usedOnOutPipe${k}`] =
          await client.tryGet(`${b}/mixerAllocation/@props/usedOnOutPipe${k}`);
      }
      rec.scalers = {};
      for (const s of SCALERS) {
        rec.scalers[s] = {
          memoryFill: await client.tryGet(`${b}/$scaler/@items/${s}/@props/memoryFill`),
          memoryCut: await client.tryGet(`${b}/$scaler/@items/${s}/@props/memoryCut`),
        };
      }
      mixers[id] = rec;
    }
    first = false;
  }

  return mixers;
}

/** Device identity. Cheap, and confirms we are talking to a LivePremier. */
export async function readIdentity(client) {
  const p = 'DeviceObject/system/$device/@items/1/@props';
  return {
    dev: await client.tryGet(`${p}/dev`),
    label: await client.tryGet(`${p}/label`),
  };
}

/**
 * The operator's own names for each screen — "Main LED", "Stream ENG" — which
 * make every other view readable. 24 cheap reads.
 *
 *   DeviceObject/$screen/@items/S<1-24>/control/@props/label
 *
 * Note it is `$screen`, not `$screenAuxGroup`: the latter has a `control` node
 * but no `label` on it (that one answers E12). Unconfigured screens return an
 * empty string, and S25 answers E12, which is how the range is known.
 *
 * These are show data — a client's screen names — so they belong in the live
 * view only. They are deliberately absent from the recorded captures in `data/`
 * and from `scripts/profile-vpu.mjs`, both of which are public.
 */
export async function readScreenNames(client, max = 24) {
  const names = {};
  for (let i = 1; i <= max; i++) {
    const label = await client.tryGet(`DeviceObject/$screen/@items/S${i}/control/@props/label`);
    if (typeof label === 'string' && label.trim()) names[`S${i}`] = label.trim();
  }
  return names;
}

File: src-tauri/src/read.rs
//! Reading the VPU map off a device — the desktop equivalent of `lib/read.js`.
//!
//! This deliberately produces the SAME JSON shape as the Node bridge's
//! `/api/vpu`, so `public/app.js` renders either without knowing which it got.
//! If you change a field here, change it there too.

use crate::awj::Awj;
use serde_json::{json, Map, Value};

const MIXER_PROPS: [&str; 8] = [
    "isEnabled",
    "cutnfillCapa",
    "usedInScreen",
    "usedInLayer",
    "channel",
    "slice",
    "capability",
    "seamlessCapa",
];

const SCREEN_STATUS_PROPS: [&str; 11] = [
    "mode",
    "mixingMode",
    "outputCount",
    "usedOutputCapabilities",
    "remainingOutputCapabilities",
    "exceedingOutputCapabilities",
    "layerCount",
    "exceedingLayerCapabilities",
    "isOptimized",
    "regionValidity",
    "isStereo3d",
];

const OUT_PIPES: u8 = 8;
const PROCESSORS: u8 = 4;
const MIXERS_PER_PROCESSOR: u8 = 16;

fn mapping(which: &str, device: &str) -> String {
    format!(
        "DeviceObject/preconfig/resources/{which}/status/mapping/$device/@items/{device}"
    )
}

/// Device identity — cheap, and confirms this is a LivePremier.
pub fn identity(a: &mut Awj) -> Value {
    let p = "DeviceObject/system/$device/@items/1/@props";
    json!({
        "dev": a.try_get(&format!("{p}/dev")),
        "label": a.try_get(&format!("{p}/label")),
    })
}

/// The operator's own screen names. `$screen`, NOT `$screenAuxGroup` — the
/// latter has a `control` node with no `label` on it.
pub fn screen_names(a: &mut Awj) -> Value {
    let mut out = Map::new();
    for i in 1..=24 {
        let path = format!("DeviceObject/$screen/@items/S{i}/control/@props/label");
        if let Some(Value::String(s)) = a.try_get(&path) {
            if !s.trim().is_empty() {
                out.insert(format!("S{i}"), Value::String(s.trim().to_string()));
            }
        }
    }
    Value::Object(out)
}

/// Per-screen resource status: spend, headroom, and whether Optimized mode is
/// on. Screens that are not configured report `DISABLED` and are skipped.
pub fn screen_status(a: &mut Awj, which: &str) -> Value {
    let mut out = Map::new();
    for i in 1..=24 {
        let base = format!(
            "DeviceObject/preconfig/resources/{which}/$screen/@items/S{i}/status/@props"
        );
        let mode = match a.try_get(&format!("{base}/mode")) {
            None => break, // past the end of the model
            Some(m) => m,
        };
        if mode.as_str() == Some("DISABLED") {
            continue;
        }
        let mut rec = Map::new();
        rec.insert("mode".into(), mode);
        for p in SCREEN_STATUS_PROPS.iter().filter(|p| **p != "mode") {
            if let Some(v) = a.try_get(&format!("{base}/{p}")) {
                rec.insert((*p).into(), v);
            }
        }
        out.insert(format!("S{i}"), Value::Object(rec));
    }
    Value::Object(out)
}

/// One side of the mapping. `None` means the device has no `$vpuMixer`
/// collection at all — the simulator behaves this way.
pub fn mapping_side(a: &mut Awj, which: &str, device: &str) -> Option<Value> {
    let mut out = Map::new();
    let mut first = true;

    for p in 1..=PROCESSORS {
        for m in 1..=MIXERS_PER_PROCESSOR {
            let id = format!("PROC_{p}_MIXER_{m}");
            let b = format!("{}/$vpuMixer/@items/{id}", mapping(which, device));

            let avail = a.try_get(&format!("{b}/@props/isAvailable"));
            let avail = match avail {
                None if first => return None, // no subtree here at all
                None => Value::Bool(false),
                Some(v) => v,
            };
            first = false;

            if avail.as_bool() != Some(true) {
                out.insert(id, json!({ "isAvailable": false }));
                continue;
            }

            let mut rec = Map::new();
            rec.insert("isAvailable".into(), Value::Bool(true));
            for prop in MIXER_PROPS.iter() {
                if let Some(v) = a.try_get(&format!("{b}/@props/{prop}")) {
                    rec.insert((*prop).into(), v);
                }
            }

            // Eight output pipes, not two — each names an output link this mixer
            // drives, and they are what makes the grid's columns real.
            let mut alloc = Map::new();
            for k in 1..=OUT_PIPES {
                let v = a
                    .try_get(&format!("{b}/mixerAllocation/@props/usedOnOutPipe{k}"))
                    .unwrap_or(Value::String("NONE".into()));
                alloc.insert(format!("usedOnOutPipe{k}"), v);
            }
            rec.insert("mixerAllocation".into(), Value::Object(alloc));

            let mut scalers = Map::new();
            for s in ["A", "B"] {
                let mut sc = Map::new();
                for f in ["memoryFill", "memoryCut"] {
                    if let Some(v) =
                        a.try_get(&format!("{b}/$scaler/@items/{s}/@props/{f}"))
                    {
                        sc.insert(f.into(), v);
                    }
                }
                scalers.insert(s.into(), Value::Object(sc));
            }
            rec.insert("scalers".into(), Value::Object(scalers));

            out.insert(id, Value::Object(rec));
        }
    }
    Some(Value::Object(out))
}
